src/main/java/com/mongodb/spark/sql/connector/read/MongoStreamPartitionReader.java

```diff
@@ -0,0 +1,204 @@
+
+/**
+ * A partition reader returned by {@link
+ * MongoStreamPartitionReaderFactory#createReader(org.apache.spark.sql.connector.read.InputPartition)}.
+ *
+ * <p>It's responsible for outputting data for a continuous stream using change streams.
+ */
+public class MongoStreamPartitionReader implements ContinuousPartitionReader<InternalRow> {
+  private static final Logger LOGGER = LoggerFactory.getLogger(MongoPartitionReader.class);
+  private static final String INVALIDATE = "invalidate";
+  private static final ResumeTokenPartitionOffset EMPTY_RESUME_TOKEN_PARTITION_OFFSET =
+      new ResumeTokenPartitionOffset(new BsonDocument());
+  private final MongoInputPartition partition;
+  private final BsonDocumentToRowConverter bsonDocumentToRowConverter;
+  private final ReadConfig readConfig;
+
+  private ResumeTokenPartitionOffset lastOffset;
+  private InternalRow currentRow;
+  private boolean closed = false;
+  private MongoClient mongoClient;
+  private MongoChangeStreamCursor<BsonDocument> changeStreamCursor;
+
+  /**
+   * Construct a new instance
+   *
+   * @param partition the partition
+   * @param bsonDocumentToRowConverter the converter from {@link BsonDocument} to {@link
+   *     InternalRow}
+   * @param readConfig the read configuration for reading from the partition
+   */
+  public MongoStreamPartitionReader(
+      final MongoInputPartition partition,
+      final BsonDocumentToRowConverter bsonDocumentToRowConverter,
+      final ReadConfig readConfig) {
+    this.partition = partition;
+    this.bsonDocumentToRowConverter = bsonDocumentToRowConverter;
+
+    this.readConfig = readConfig;
+    this.currentRow = null;
+    this.lastOffset =
+        new ResumeTokenPartitionOffset(partition.getResumeTokenOffset().getResumeToken());
+
+    LOGGER.debug(
+        "Creating partition reader for: PartitionId: {} with Schema: {}",
+        partition.getPartitionId(),
+        bsonDocumentToRowConverter.getSchema());
+  }
+
+  @Override
+  public PartitionOffset getOffset() {
+    Assertions.ensureState(
+        () -> !closed, () -> "Cannot call getOffset() on a closed PartitionReader.");
+    LOGGER.trace("getOffset called, returning: {}", lastOffset);
+    return lastOffset;
+  }
+
+  @Override
+  public boolean next() {
+    Assertions.ensureState(() -> !closed, () -> "Cannot call next() on a closed PartitionReader.");
+    return withCursor(
+        c -> {
+          try {
+            BsonDocument next = c.next();
+            lastOffset = new ResumeTokenPartitionOffset(c.getResumeToken());
+            if (next.getString("operationType", new BsonString(""))
+                .getValue()
+                .toLowerCase(Locale.ROOT)
+                .equals(INVALIDATE)) {
+              LOGGER.info(
+                  "Change stream cursor has been invalidated. This happens when a collection is dropped. Closing cursor.");
+              lastOffset = EMPTY_RESUME_TOKEN_PARTITION_OFFSET;
```

Comment:
So this is tricky. An invalidated change stream event doesn't mean there wont be more changes on the collection in the future. Its just that invalidate events happen to kill the current cursor and requires a new cursor.

Why use an empty resume token? this is a hang over from pre MongoDB 4.2 where users would have to use resumeAfter - however, if we limit users to MongoDB 4.2+ then startAfter is the best api. So I have updated to use that. It will stop a data race and potential loss of data.